### DISCUSSION TITLE
How to sent multiple, different kafka messages in one thread group

### DISCUSSION THREAD

**Discussion by poorSock** [sngular/kloadgen#351] [General]
Hi there,

my use case is as follows:

- I want to sent a group of kafka messages (like, a lot of, but for simplicity let's say "2" here, should not make any difference)
- these are different messages which may look completely different

To achieve this:

- I created a thread group
- created a KLG Value Schema Serializer Config + a Java request
- created another KLG Value Schema Serialer Config

Now I validated the thread group.
Ok I get only 1 message, which is logical, as I have only added 1 sampler. So I added another java request and sorted it like this:

- config1
- request
- config2
- request

now I get 2 messages but now both requests seem to use config1 and sent the same data
no matter how I order this, or do it, I get the data from config1, config2 is ignored

I want to sent multiple requests with multiple configs in 1 thread group, is this possible with KLG?

**Why don't you use multiple thread groups?**
Yes, I am aware that I can use multiple thread groups, with 1 config and 1 request in each thread group and get the desired messages out of it. However, as I stated my group of kafka messages reflects 1 testcase.
Lets say I would loop over each thread group 100000 times. Then I wouldn't be sure anymore, that the messages belonging to 1 testcase arrive next to each other, or even close to each other

**Why don't you use 1 thread group with only 1 config and 1 request and use variables to fill in the data?**
Yes I could possible do this. But because I have groups of events and in each group all events look quite different I would have to write ugly code in JMeter. Also I would ask myself the question why I would need this plugin then, or JMeter in the first place.

**Comment by dhergonsngular** [comment]
Hi @poorSock,

If I understand correctly, you want to apply different configs for each request within the same thread group.

As you commented, KLG processes each request, but it does not follow the sequential config+request order you are trying to implement. 
For this you can add in JMeter simple controllers and in each of them add the config+request.
KLG will process each request with the configuration of the created group.

![image](https://user-images.githubusercontent.com/107672198/225231796-404df306-a250-4040-8133-3c1e3bbc8607.png)

Attached is the modified example-test-plan, can you check if it is the solution you are looking for?

[Example-Test-Plan.jmx.zip](https://github.com/sngular/kloadgen/files/10976882/Example-Test-Plan.jmx.zip)

**Reply by poorSock** [reply]
Hi,
was exactly what I wanted. Clearly lacking JMeter experience here, Thank you!
